### PR TITLE
[WIP] [GPT] Add MP enhancement and accuracy check scripts.

### PR DESCRIPTION
**Pull request by ZHUI**
### PR types

Bug fixes 
### PR changes

Others
### Description

[GPT] Add MP enhancement and accuracy check scripts.
[WIP] Don't Merge!

### CODE REVIEW

File: examples/language_model/gpt/run_pretrain_static.py [hunk 1]
```diff
@@ -236,13 +243,25 @@ def do_train(args):
                     data_holders=data_holders,
                     pipeline_mode=False, )
 
+                if args.model_name_or_path in pretrained_models_list:
+                    model_config = model_class.pretrained_init_configuration[
+                        args.model_name_or_path]
+                    if model_config["vocab_size"] % 8 != 0:
```
Comment (ZeyuChen) [inline examples/language_model/gpt/run_pretrain_static.py]: 这么tricky的%8操作，最好加下注释

File: examples/language_model/gpt/run_pretrain_static.py [hunk 2]
```diff
@@ -83,6 +83,8 @@ def dist_optimizer(args, topo):
     dist_strategy.recompute = args.use_recompute
     dist_strategy.pipeline = args.pp_degree > 1
 
+    #dist_strategy.without_graph_optimization = True
```
Comment (wangxicoding) [inline examples/language_model/gpt/run_pretrain_static.py]: 这个可以先帮我删掉，自己测试用的

File: paddlenlp/ops/distributed/parallel.py [hunk 3]
```diff
@@ -69,69 +130,65 @@ def __init__(self,
...
-
-        x_shard = paddle.shard_index(x, self.origin_num_embeddings,
-                                     self.num_partitions, self.inner_rank,
-                                     self.per_part_embeddings - 1)
-        if len(origin_input_shape) == 2:
-            x_shard = paddle.squeeze(x_shard, axis=-1)
-
-        emb_out = self.embedding(x_shard)
+        # TODO(wangxi): dynamic group
+        group = None
```
Comment (wangxicoding) [inline paddlenlp/ops/distributed/parallel.py:174]: 这个需要先支持动态图group吗，还是等我动静统一的API。当前group=None动态图只能跑MP，不能跑混合并行

File: paddlenlp/transformers/gpt/modeling.py [hunk 4]
```diff
@@ -425,22 +460,22 @@ def __init__(self,
...
         super(GPTEmbeddings, self).__init__()
-        # TODO @ZHUI Use ParallelEmbedding
-        #if topo is None or topo.mp_info.size == 1:
-        self.word_embeddings = nn.Embedding(
-            vocab_size,
-            hidden_size,
-            weight_attr=paddle.ParamAttr(
-                name="word_embeddings",
-                initializer=nn.initializer.Normal(
+        #if True:
```
Comment (wangxicoding) [inline paddlenlp/transformers/gpt/modeling.py]: 调试用的，可以删掉

File: paddlenlp/transformers/gpt/modeling.py [hunk 5]
```diff
@@ -750,6 +785,31 @@ def __init__(self, gpt):
         self.gpt = gpt
         self.apply(self.init_weights)
 
+    def parallel_matmul(self, lm_output, logit_weights, parallel_output, topo):
+        #hcg = fleet.get_hybrid_communicate_group()
```
Comment (wangxicoding) [inline paddlenlp/transformers/gpt/modeling.py]: 可以删掉，或者加上group支持动态图

File: paddlenlp/transformers/gpt/modeling.py [hunk 6]
```diff
@@ -767,10 +827,13 @@ def forward(self,
         else:
             encoder_outputs = outputs
         # TODO @ZHUI Use all_to_all to
-        logits = paddle.matmul(
-            encoder_outputs,
-            self.gpt.embeddings.word_embeddings.weight,
-            transpose_y=True)
+        # logits = paddle.matmul(
```
Comment (wangxicoding) [inline paddlenlp/transformers/gpt/modeling.py]: 可删

File: paddlenlp/ops/distributed/parallel.py (hunk 3 above)
Comment (ZHUI) [reply paddlenlp/ops/distributed/parallel.py:174]: 先不支持 动态图 吧。动静统一了再另外 pr 吧

File: examples/language_model/gpt/run_pretrain_static.py (hunk 1 above)
Comment (ZHUI) [reply examples/language_model/gpt/run_pretrain_static.py]: 好的，这段干脆删除了。这个主要是调整vocab_size，方便切分参数。

File: examples/language_model/gpt/run_pretrain_static.py (hunk 2 above)
Comment (ZHUI) [reply examples/language_model/gpt/run_pretrain_static.py]: 已删除

File: paddlenlp/transformers/gpt/modeling.py (hunk 4 above)
Comment (ZHUI) [reply paddlenlp/transformers/gpt/modeling.py]: 已删除

File: paddlenlp/transformers/gpt/modeling.py (hunk 5 above)
Comment (ZHUI) [reply paddlenlp/transformers/gpt/modeling.py]: 已删除

File: paddlenlp/transformers/gpt/modeling.py (hunk 6 above)
Comment (ZHUI) [reply paddlenlp/transformers/gpt/modeling.py]: 已删除

**Review by wangxicoding — DISMISSED** [review]
LGTM

File: examples/language_model/gpt/run_pretrain_static.py [hunk 7]
```diff
@@ -204,6 +211,10 @@ def do_train(args):
 
     # Define the input data in the static mode
 
+    model_class, tokenizer_class = MODEL_CLASSES[args.model_type]
```
Comment (ZeyuChen) [inline examples/language_model/gpt/run_pretrain_static.py:214]: 这里选择model_type还有什么可以选呢？是固定GPT好还是提供选择好？

Comment (ZHUI) [reply examples/language_model/gpt/run_pretrain_static.py:214]: 这里是中英文可选，后续需要训练中文模型的话，还是可选比较好。

File: examples/language_model/gpt/args.py [hunk 8]
```diff
@@ -217,10 +219,25 @@ def parse_args(MODEL_CLASSES):
...
+        help="The attention probs dropout prob.")
     # Other config
     parser.add_argument(
         "--seed", type=int, default=1234, help="Random seed for initialization")
+    parser.add_argument(
+        "--check_accuracy",
+        type=str2bool,
+        nargs='?',
+        const=False,
+        help="check accuracy for training process.")
```
Comment (wawltor) [inline examples/language_model/gpt/args.py]: check->Check

File: examples/language_model/gpt/run_pretrain.py [hunk 9]
```diff
@@ -144,11 +150,15 @@ def do_train(args):
...
-        opt_dict = paddle.load(
-            os.path.join(args.model_name_or_path, "model_state.pdopt"))
-        optimizer.set_state_dict(opt_dict)
+        logger.info("Try to load checkpoint from %s " % args.model_name_or_path)
+        opt_path = os.path.join(args.model_name_or_path, "model_state.pdopt")
+        if os.path.exists(opt_path):
+            opt_dict = paddle.load(opt_path)
+            optimizer.set_state_dict(opt_dict)
+        else:
+            logger.warning("No checkpoint file found in %s." % opt_path)
```
Comment (wawltor) [inline examples/language_model/gpt/run_pretrain.py]: 这里应该要表明的load信息是optimizer吧

File: examples/language_model/gpt/run_pretrain_static.py [hunk 10]
```diff
@@ -236,13 +240,32 @@ def do_train(args):
...
+
+                    model_config[
+                        "hidden_dropout_prob"] = args.hidden_dropout_prob
+                    model_config[
+                        "attention_probs_dropout_prob"] = args.attention_probs_dropout_prob
+                    model_config["topo"] = topo
+
+                    model = guard(f'gpu:{args.pp_degree -1}')(
+                        GPTForPretraining)(guard(f'gpu:0')(GPTModel)(
+                            **model_config))
```
Comment (wawltor) [inline examples/language_model/gpt/run_pretrain_static.py:255]: Todo 删除静态图Pipeline

**Review by wawltor — APPROVED** [review]
LGTM
